Dependency graph parametrization

istalker2 opened the pull request:
* It is possible to declare parameters of the flows with description and default value
* Argument values can be used anywhere in the resource definitions except for the container parts
  (not to cause conflicts with shell variables)
* Dependencies can carry arguments between resources (for example, to map self arguments to parameters
  of the flow being triggered)
* It is possible to have several dependencies on the same object with different arguments. If the argument is
  used as part of the resource name it will produce different resource for each such dependency despite
  both parent and child being the same
* Arguments can be specified in the command line (kubeac run myFlow --arg arg1=value1, --arg arg2: value2
   Argument are validated: undeclared arguments are not allowed (unless there is a --undeclared-args flag),
   arguments without default values must be provided

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mirantis/k8s-appcontroller/224)
<!-- Reviewable:end -->

jellonek commented:
One minor nit and docstrings as usual... Rest <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 51 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[cmd/deploy.go, line 122 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-KfqYGtUnEpU473sR0dT:-KfqYGtVf05w4wtjhN2G:bg5zk40) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/55f85182e36646a11eef38ef68f9a2d45f4f9bf1/cmd/deploy.go#L122)):*
> ```Go
> 	res := argRe.FindStringSubmatch(arg)
> 	if len(res) != 3 {
> 		return "", "", fmt.Errorf("Invalid argument %s", arg)
> ```

Please start error with lowercase.

---

*[pkg/client/client.go, line 182 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-KfqY_lbHWOOUiHFbeKd:-KfqY_lbHWOOUiHFbeKe:bmzrj4l) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/55f85182e36646a11eef38ef68f9a2d45f4f9bf1/pkg/client/client.go#L182)):*
> ```Go
> }
> 
> // Returns AC namespace
> ```

https://golang.org/doc/effective_go.html#commentary there is sentence:
"If every doc comment begins with the name of the item it describes, the output of godoc can usefully be run through grep."

This is why we are trying to keep thing name as a start of docstring for that thing (no mater if it's type, variable or func).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-:-KfqaMQIyDfjwwrpkhbo:b-96myuu)*
<!-- Sent from Reviewable.io -->

nebril commented:
Reviewed 18 of 51 files at r1, 11 of 12 files at r2.
Review status: 28 of 29 files reviewed at latest revision, 8 unresolved discussions.

---

*[cmd/deploy.go, line 55 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-JCNevANUPWDnLXBC:-Kg-JCNevANUPWDnLXBD:b-7iun5x) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L55)):*
> ```Go
> 	}
> 
> 	flowArgs, err := cmd.Flags().GetStringArray("arg")
> ```

Is argument to the cli called `arg`? If yes, I suggest changing it to reflect what this argument really is.

---

*[cmd/deploy.go, line 72 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-JnwAigowRLsTpGfK:-Kg-JnwAigowRLsTpGfL:b3cpt0i) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L72)):*
> ```Go
> 
> 	if len(args) > 1 {
> 		log.Fatal("Too many command line arguments")
> ```

I think this should be handled by `pflag` lib.

---

*[cmd/deploy.go, line 112 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-KEBWXoQtUs9NBDyM:-Kg-KEBWXoQtUs9NBDyN:b6eerc6) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L112)):*
> ```Go
> func parseArg(arg string) (string, string, error) {
> 	res := argRe.FindStringSubmatch(arg)
> 	if len(res) != 3 {
> ```

Please document or preferably test logic of this function.

---

*[cmd/deploy.go, line 141 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-KXEt19wZ0X17GMx8:-Kg-KXEt19wZ0X17GMx9:botruo8) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L141)):*
> ```Go
> 
> 	var args []string
> 	run.Flags().StringArrayVar(&args, "arg", []string{}, "Flow arguments (key=value)")
> ```

`arg` is not a proper name for this argument. Also, how do I provide several arguments? Please include this information in help string.x

---

*[pkg/client/client.go, line 182 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-KfqY_lbHWOOUiHFbeKd:-Kg-LASKB3x9G6TtM0ge:b32v57) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/55f85182e36646a11eef38ef68f9a2d45f4f9bf1/pkg/client/client.go#L182)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

https://golang.org/doc/effective_go.html#commentary there is sentence:
"If every doc comment begins with the name of the item it describes, the output of godoc can usefully be run through grep."

This is why we are trying to keep thing name as a start of docstring for that thing (no mater if it's type, variable or func).
</blockquote></details>

+1

---

*[pkg/copier/copier.go, line 16 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-M-ynANei8NPyW-es:-Kg-M-ynANei8NPyW-et:b-d2bhb4) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/pkg/copier/copier.go#L16)):*
> ```Go
> 
> // Based on the work of Joel Scoble:
> // https://github.com/mohae/deepcopy/blob/master/deepcopy.go
> ```

Why not add this repo to dependencies and use the copy function wrapped in your own functions?

---

*[pkg/resources/daemonset.go, line 54 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-N2rsjQXYIGElRW5Z:-Kg-N2rtR1s2CzgHD-cI:b-dov593) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/pkg/resources/daemonset.go#L54)):*
> ```Go
> func (d daemonSetTemplateFactory) New(def client.ResourceDefinition, c client.Interface, gc interfaces.GraphContext) interfaces.Resource {
> 	newDaemonSet := parametrizeResource(def.DaemonSet, gc,
> 		"Spec.Template.Spec.Containers.Env",
> ```

Why only replacing the values in Env? The commend stands for this and most other resources.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224)*
<!-- Sent from Reviewable.io -->

jellonek commented:
Review status: 28 of 29 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/copier/copier.go, line 16 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-M-ynANei8NPyW-es:-Kg-PohN0bcyMa7atkzg:bhq0uu4) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/pkg/copier/copier.go#L16)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

Why not add this repo to dependencies and use the copy function wrapped in your own functions?
</blockquote></details>

PTAL into linked file. IMO there is no way to reuse these funcs with needs which are added there (replacing templates).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224)*
<!-- Sent from Reviewable.io -->

nebril commented:
Review status: 28 of 29 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/copier/copier.go, line 16 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-M-ynANei8NPyW-es:-Kg-REaFWV_WpVX0P0Vl:b7e4hsb) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/pkg/copier/copier.go#L16)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

PTAL into linked file. IMO there is no way to reuse these funcs with needs which are added there (replacing templates).
</blockquote></details>

Makes sense, thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224)*
<!-- Sent from Reviewable.io -->

istalker2 commented:
Review status: 28 of 29 files reviewed at latest revision, 8 unresolved discussions.

---

*[cmd/deploy.go, line 122 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-KfqYGtUnEpU473sR0dT:-Kg-rbcg4deptObrPRua:b-7ql1td) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/55f85182e36646a11eef38ef68f9a2d45f4f9bf1/cmd/deploy.go#L122)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Please start error with lowercase.
</blockquote></details>

fixed

---

*[cmd/deploy.go, line 55 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-JCNevANUPWDnLXBC:-Kg-qNPmQcRQnomo_l4D:b-oncol3) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L55)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

Is argument to the cli called `arg`? If yes, I suggest changing it to reflect what this argument really is.
</blockquote></details>

Change to what? This is user-defined parameters, not something predefined. You say "kubeac run myFlow --arg myParameter1=myValue1 --arg myParameter2=myValue2"

---

*[cmd/deploy.go, line 72 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-JnwAigowRLsTpGfK:-Kg-qgyH8gDoE0cYUIo7:b-rbvbpk) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L72)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

I think this should be handled by `pflag` lib.
</blockquote></details>

This doesn't belong in this review as this is a line from PR #223. It validates that you can't say "kubeac run myFlow1 myFlow2" i.e. that there can be at most one flow in the command line.

---

*[cmd/deploy.go, line 112 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-KEBWXoQtUs9NBDyM:-Kg-r7FyEaqZXkiInKUY:b-hgf30v) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L112)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

Please document or preferably test logic of this function.
</blockquote></details>

There will be tests. The logic is to parse expressions like "myArg=myValue" or "myArg: myValue" etc to have the flexibility. The function just splits the flag value by any delimiter with regexp, and trims spaces

---

*[cmd/deploy.go, line 141 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-KXEt19wZ0X17GMx8:-Kg-rg8jKCOkHE9k3pfT:bom0652) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/cmd/deploy.go#L141)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

`arg` is not a proper name for this argument. Also, how do I provide several arguments? Please include this information in help string.x
</blockquote></details>

This is StringArrayVar i.e. list of arguments. The library automatically generates help saying that this is a list. You provie multiple values with --arg arg1=value1 --arg arg2=value2

---

*[pkg/client/client.go, line 182 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-KfqY_lbHWOOUiHFbeKd:-Kg-ryOZF08Xz0QM6AX5:b-7ql1td) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/55f85182e36646a11eef38ef68f9a2d45f4f9bf1/pkg/client/client.go#L182)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

+1
</blockquote></details>

fixed

---

*[pkg/resources/daemonset.go, line 54 at r2](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kg-N2rsjQXYIGElRW5Z:-Kg-s8L_quU_m6AZC9Ju:b-vs0z88) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/88997b8fe5b691053d03716aec4ade72379aef7a/pkg/resources/daemonset.go#L54)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary><blockquote>

Why only replacing the values in Env? The commend stands for this and most other resources.
</blockquote></details>

The function automatically adds ObjectMeta to this list. Currently it is enough for our needs because the main purpose was to use parameters in resource names and pass those values to bash scripts through env. If you think of some other places where it will be useful please post it here. I just don't have any in my head and we always can add something to this list later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224)*
<!-- Sent from Reviewable.io -->

jellonek commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 12 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-:-KgmwKOxCJ48HU_nk1jO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nebril commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 28 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-:-Kh1bwgA5PYQne_ZSLUl:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nebril commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 27 of 28 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-:-Kh2a6m9jt6H2xDx-1jL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jellonek commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/client/flows.go, line 43 at r3](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-Kh6JPBMmEfRQzywl8-0:-Kh6JPBMmEfRQzywl8-1:bcnghd0) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/2b8d23ecf60cab32db35ac42cbefd7a1e885178d/pkg/client/flows.go#L43)):*
> ```Go
> }
> 
> type FlowParameter struct {
> ```

Missing docstring.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/224#-:-Kh6JWHJ8jvWUgfgCGrq:bnfp4nl)*
<!-- Sent from Reviewable.io -->
